### PR TITLE
🗣️ fix: Keep Provider Error Text on Unclassified Agent Failures

### LINKED ISSUE #16032: [Bug]: Provider error messages are discarded since v0.8.8-rc3, contradicting the comment in errors.ts
### What happened?

Since v0.8.8-rc3, error messages returned by an upstream provider are no
longer shown to the user. They are replaced by the generic
`UPSTREAM_MODEL_ERROR_FALLBACK` string with a status code. Under v0.8.7
the provider message was appended and reached the user.

This contradicts a comment in the codebase. `packages/api/src/agents/errors.ts`
(lines 70-71) states about codes that are not listed:

> Codes absent here keep the provider's own message (minus the docs URL),
> which is more specific than any generic string we could write.

But `packages/api/src/agents/failures/terminal.ts` (lines 76-92) discards
the provider message when classification fails and substitutes
`UPSTREAM_MODEL_ERROR_FALLBACK` instead.

Classification only recognises two cases (`errors.ts:73-76`):

```js
const LANGCHAIN_ERROR_TYPES = {
  MODEL_NOT_FOUND:  ErrorTypes.MODEL_NOT_FOUND,
  MODEL_RATE_LIMIT: ErrorTypes.MODEL_RATE_LIMIT,
};
```

and it keys off `error.lc_error_code`, a field set by LangChain
(`errors.ts:82-91`). Neither the HTTP status nor `error.type` or
`error.message` of the OpenAI/Anthropic error envelope is examined. A
compliant `invalid_request_error` with a `message` is therefore never read.

### Why this matters

This affects anyone running a gateway, proxy or OpenAI-compatible endpoint
in front of LibreChat. In our deployment a privacy proxy sits between
LibreChat and Anthropic and rejects requests it cannot mask safely,
returning a 400 with an explanatory message. Under v0.8.7 the user saw
that explanation. Under rc3 they see only
"The model provider could not complete this request (status 400)", with
no way to tell a privacy rejection from an outage.

The rejection itself still works. Only the explanation is lost.

### Version Information

LibreChat v0.8.8-rc3 (1df44848), compared against v0.8.7 (9e74cc0e)
Node.js v24.16.0
Native install on Windows Server (no Docker)

### Steps to Reproduce

1. Configure a custom endpoint that returns an OpenAI-style error, e.g.
   `{"error":{"code":400,"message":"<your text>","type":"invalid_request_error"}}`
2. Send a message through an agent using that endpoint
3. The user sees the generic fallback, not `<your text>`

Under v0.8.7 the same setup showed the provider text, appended by
`api/server/controllers/agents/client.js`:

```js
[ContentTypes.ERROR]: `An error occurred while processing the request${err?.message ? `: ${err.message}` : ''}`
```

### Expected behavior

As stated in the comment in `errors.ts`: unclassified errors keep the
provider's own message.

### Additional notes

`getUserFacingProviderError` (`errors.ts:107-115`) does pass the message
through, but it is only reached from the OpenAI-compatible ingress routes,
not from the UI path (`chat.js:23` -> `controllers/agents/request` ->
`client.js`, which uses `terminal.ts` exclusively).

The `fallback` argument of `getUserFacingError` does not help either: it
only applies when `upstreamModelError == null`, and a provider 400 is
registered as an upstream model error.

The comment dates from 982a9a514 (#15421, 2026-08-31), the discarding code
from 19eb24921 (#15837, 2026-09-12). The later change does not appear to
have been aware of the earlier comment.

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```shell
{"errorCode":"UPSTREAM_MODEL_ERROR","errorOrigin":"model_provider","errorType":"400","level":"error","message":"[api/server/controllers/agents/client.js #sendCompletion] Upstream model error","requestId":"<redacted>","request_id":"<redacted>","request_method":"POST","request_path":"/api/agents","status":400,"timestamp":"2026-09-16T19:24:57.477Z","traceId":"<redacted>","type":"Error","userId":"<redacted>"}

The failure is recorded with code, origin, type and status, but the
provider's own message is absent here as well - so neither the user nor
the operator can tell what the upstream actually said.
```

### Screenshots

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by berry-13**
## Summary

I restored provider explanations on unclassified agent failures. Since #15837, a gateway or OpenAI-compatible provider rejection could show only a generic status headline, discarding the reason from both the error card and transcript.

- Preserve provider wording beneath the localized headline when content protection is explicitly disabled.
- Withhold provider text when a content policy is active or an older caller omits the protection setting.
- Retain at most `endpoints.agents.maxProviderErrorChars` characters, defaulting to 2,000; allow 0 to suppress detail and validate the range through 1,000,000.
- Bound provider text before scanning it, handle hostile message accessors, and remove LangChain troubleshooting URLs with bounded lookahead.
- Parse quoted braces and escaped characters correctly so provider messages remain readable instead of exposing the serialized error payload.
- Reuse accessible inline/disclosure rendering and existing localized strings.

Fixes #16032

## How it works

Each agent ingress passes the loaded content policy and retention limit into the terminal observer. Log metadata retains its existing redaction; the user-facing error carries the gated explanation.

```text
chatCompletion / resumeCompletion / OpenAI / Responses
  createTerminalRunErrorObserver(protectionEnabled, maxProviderErrorChars)
    getProviderErrorMessage → bounded, sanitized text
    persisted upstream_model_error payload → localized ModelError
      ErrorWithDetail → inline explanation or expandable details
```

An omitted protection setting preserves the previous safe behavior for JavaScript callers. Existing error payloads without a message still render their localized headline. The YAML example documents the new retention setting.

## Change Type

- [x] Bug fix (non-breaking change which fixes an issue)

## Testing

Focused checks on the completed implementation:

- `packages/api`: `errors.spec.ts` and `failures/terminal.spec.ts` — 63 tests passed, including omitted protection, hostile accessors, configurable limits, bounded scanning, and troubleshooting URLs crossing the output boundary.
- `api`: `server/controllers/agents/client.test.js` — 254 tests passed, including resumed failures with default/custom limits and protected payloads.
- `client`: `utils/json.spec.ts` and `Messages/Content/__tests__/Error.spec.tsx` — 145 tests passed, including quoted braces, escaped quotes/backslashes, localized status, inline details, and disclosure accessibility.
- `packages/data-provider`: `config.spec.ts` and `provider.spec.ts` — 214 tests passed, including default/custom limits and invalid configuration.
- `npx tsc --noEmit` passed in `packages/api`, `packages/data-provider`, and `client`.
- Focused ESLint, whitespace checks, and API/data-provider package builds passed.

- Mock browser tests: completion failure visibility and error-response follow-up persistence — 2 passed against the real backend and Chrome.
- `E2E_CHROMIUM_CHANNEL=chrome npm run lighthouse` passed with 250 ms injected database latency: median LCP 3,514 ms (4,500 ms budget), CLS 0.017 (0.1), TBT 23 ms (500 ms).

### **Test Configuration**:

Unprotected deployments retain bounded explanations. Protected deployments and callers omitting the protection flag retain only the generic headline/status. Regressions cover both shapes, limits below/above the default, and zero to hide details.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

Please review the current PR head 9f6298496e088f20b68ec84188be77cc980db8bb. Confirm that this exact commit is the reviewed commit and ignore findings that apply only to earlier heads.

File: packages/api/src/agents/failures/terminal.ts [hunk 1]
```diff
@@ -86,9 +93,16 @@ export function createTerminalRunErrorObserver({
       }
 
       const { status } = getSafeErrorMetadata(upstreamModelError);
+      /** Unclassified: the provider's own explanation is the only account of what happened, and a
+       *  rejection from a gateway or proxy carries it as the whole point of the 400. The status
+       *  headlines it either way, so a deployment withholding provider text loses no taxonomy. */
+      const providerMessage = protectionEnabled
+        ? undefined
+        : (getProviderErrorMessage(upstreamModelError) ?? getProviderErrorMessage(error));
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/failures/terminal.ts]: Fixed in 9f6298496e088f20b68ec84188be77cc980db8bb. Provider detail is exposed only when protectionEnabled is explicitly false; omitted options preserve redaction for older JavaScript consumers. The terminal observer regression exercises both true and undefined and verifies the persisted payload excludes submitted content.

File: packages/api/src/agents/failures/terminal.ts [hunk 2]
```diff
@@ -86,9 +93,16 @@ export function createTerminalRunErrorObserver({
...
       return `${UPSTREAM_MODEL_ERROR_FALLBACK}\n${JSON.stringify({
         type: ErrorTypes.UPSTREAM_MODEL_ERROR,
         ...(status != null ? { status } : {}),
+        ...(providerMessage != null ? { message: providerMessage } : {}),
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/failures/terminal.ts:110]: Fixed in 9f6298496e088f20b68ec84188be77cc980db8bb. extractJson now tracks quoted strings and escapes, and ignores closing braces before the object starts. Parser tests cover nested objects, quoted braces, escaped quotes and backslashes; renderer tests exercise the actual headline-plus-JSON wire payload and verify localized status plus provider detail.

File: packages/api/src/agents/errors.ts [hunk 3]
```diff
@@ -114,6 +114,27 @@ export function getUserFacingProviderError(error: unknown, protectionEnabled: bo
   return stripLangChainTroubleshootingUrl(error.message) || GENERIC_PROVIDER_ERROR;
 }
 
+/** Longest provider text a failure carries to a reader; error bodies are unbounded. */
+const MAX_PROVIDER_ERROR_TEXT = 2000;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/errors.ts]: Fixed in 9f6298496e088f20b68ec84188be77cc980db8bb. endpoints.agents.maxProviderErrorChars is validated by configSchema, defaults to 2000, and is passed through all four agent entry points. Zero suppresses detail. Config, observer and resumed-controller tests verify defaults and overrides. Documented in librechat.example.yaml.

File: packages/api/src/agents/errors.ts [hunk 4]
```diff
@@ -114,6 +114,27 @@ export function getUserFacingProviderError(error: unknown, protectionEnabled: bo
...
+ * than a string, and the docs URL LangChain stamps in is not for a reader.
+ */
+export function getProviderErrorMessage(error: unknown): string | undefined {
+  const raw =
+    error != null && typeof error === 'object' ? readErrorProperty(error, 'message') : error;
+  if (typeof raw !== 'string') {
+    return undefined;
+  }
+  const message = stripLangChainTroubleshootingUrl(raw).trim();
+  return message.length === 0 ? undefined : message.slice(0, MAX_PROVIDER_ERROR_TEXT);
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/agents/errors.ts]: Fixed in 9f6298496e088f20b68ec84188be77cc980db8bb. The input is sliced to the configured output limit plus 256 characters of bounded troubleshooting-URL lookahead before stripping. A multi-megabyte regression spies on the real string scanner to verify its input stays bounded; another test covers a troubleshooting suffix crossing the output boundary.
